File: python/contains_duplicate_ii.py
class Solution:
    def containsNearbyDuplicate(self, nums: list[int], k: int) -> bool:
        l: int = 0
        numsSet: set[int] = set()
        for r in range(len(nums)):
            if r - l > k:
                numsSet.remove(nums[l])
                l += 1
            if nums[r] in numsSet:
                return True

            numsSet.add(nums[r])
        return False


def test_containsNearbyDuplicate_case_1():
    # arrange
    nums: list[int] = [1, 2, 3, 1]
    k: int = 3
    expected: bool = True

    # act
    solution = Solution()
    actual = solution.containsNearbyDuplicate(nums, k)

    # assert
    assert expected == actual


def test_containsNearbyDuplicate_case_2():
    # arrange
    nums: list[int] = [1, 0, 1, 1]
    k: int = 1
    expected: bool = True

    # act
    solution = Solution()
    actual = solution.containsNearbyDuplicate(nums, k)

    # assert
    assert expected == actual


def test_containsNearbyDuplicate_case_3():
    # arrange
    nums: list[int] = [1, 2, 3, 1, 2, 3]
    k: int = 2
    expected: bool = False

    # act
    solution = Solution()
    actual = solution.containsNearbyDuplicate(nums, k)

    # assert
    assert expected == actual

File: rust/src/code/contains_duplicate_ii.rs
struct Solution {}

impl Solution {
    pub fn contains_nearby_duplicate(nums: Vec<i32>, k: i32) -> bool {
        use std::collections::HashSet;

        let mut l: usize = 0;
        let mut nums_set: HashSet<i32> = HashSet::new();
        for r in 0..nums.len() {
            if r - l > k as usize {
                nums_set.remove(&nums[l]);
                l += 1
            }
            if nums_set.contains(&nums[r]) {
                return true;
            }
            nums_set.insert(nums[r]);
        }
        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_contains_nearby_duplicate_case_1() {
        // arrange
        let nums: Vec<i32> = vec![1, 2, 3, 1];
        let k: i32 = 3;
        let expected: bool = true;

        // act
        let actual = Solution::contains_nearby_duplicate(nums, k);

        // assert
        assert_eq!(expected, actual);
    }
}
